### ISSUE TITLE
Architecture design improvement

### ISSUE THREAD

**Issue by CyborTronik** [aaberg/sql2o#209]
Can we change the API let's say for version 2.0?
This is a totally great library that I found useful for many reasons. I'm trying to use it in another opensource project.
But there is a serious lack of architecture design(BTW there are too many open source projects may deal with exact the same issues).
First of all the project must meet S.O.L.I.D. principles that will make it more extendable and flexible.
For now I would like to highlight that I'm missing at least Single responsibility principle (SRP).

For example I would like to build a Query and then just execute it, at another level or in other class, maybe thread(what if I need a queue of queries?).

So current issues that the Query do at least few things: actually is query builder, then its a Query executor and finally a result mapper.
So is a total violation of SRP and that brakes usability which is most important part of modern libraries and apps.

Thanks

**Comment by aaberg** [comment]
I totally agree! When I started writing the code on sql2o I didn't have a clear plan and roadmap, I just had a vague idea. It has gradually evolved and grown into what it is today.
Meanwhile a lot of people have contributed, and added their touch. If not for the contributors, sql2o would never have become what it is today

**Comment by aaberg** [comment]
I totally agree,

When I started writing the code for sql2o, I didn't have a clear plan and roadmap, I just had a vague idea of a small library that would make things easier when accessing databases. The first version was really nothing more than a proof of concept I made for a project I was working on. I had not planned for it to be used by a lot of people. That is the root cause of the 'missing' architecture design.

Along the way a lot of people have added their touch to the framework, which really have made it a lot more powerful than it otherwise would have been. Especially @dimzon and @aldenquimby has put in a lot of effort, and has helped me to refactor some of the sql2o core, and added more structure than there originally was. 

But, the root problem really is, that original product was never intended to what is today. It has been discussed before to change the API ( #83 ). 

That brings me to another challenge, which is *time*. A new version with updated API would require us to rewrite and refactor most of the internals of sql2o, which undoubtedly will require a lot of time and effort. That is often a challenge in open source projects, since they are based on work done in free time.
